src/boot/efi/boot.c

```diff
@@ -2337,6 +2357,46 @@ EFI_STATUS efi_main(EFI_HANDLE image, EFI_SYSTEM_TABLE *sys_table) {
 
         config_load_defaults(&config, root_dir);
 
+        if (setup_mode_enabled() && config.auto_enroll) {
+                Print(L"Secure Boot auto enrollement is enabled.\n");
+
+                /* Read the required files in order to perform auto-enrollement */
+                err_sb_db  = file_read(root_dir, L"\\db.auth", 0, 0, &sb_db_buffer, &sb_db_size);
+                err_sb_kek = file_read(root_dir, L"\\KEK.auth", 0, 0, &sb_kek_buffer, &sb_kek_size);
+                err_sb_pk  = file_read(root_dir, L"\\PK.auth", 0, 0, &sb_pk_buffer, &sb_pk_size);
+
+                if (!EFI_ERROR(err_sb_db) && !EFI_ERROR(err_sb_kek) && !EFI_ERROR(err_sb_pk)) {
+                        err = efivar_set_raw(EFI_SIG_DB_GUID, L"db", sb_db_buffer, sb_db_size, sb_vars_options);
+                        if (EFI_ERROR(err)) {
+                                Print(L"Failed to write db secure boot variable: %r\n", err);
+                                uefi_call_wrapper(BS->Stall, 1, 3 * 1000 * 1000);
+                                return err;
+                        }
+
+                        err = efivar_set_raw(EFI_GLOBAL_GUID, L"KEK", sb_kek_buffer, sb_kek_size, sb_vars_options);
+                        if (EFI_ERROR(err)) {
+                                Print(L"Failed to write db secure boot variable: %r\n", err);
+                                uefi_call_wrapper(BS->Stall, 1, 3 * 1000 * 1000);
+                                return err;
+                        }
+
+                        err = efivar_set_raw(EFI_GLOBAL_GUID, L"PK", sb_pk_buffer, sb_pk_size, sb_vars_options);
+                        if (EFI_ERROR(err)) {
+                                Print(L"Failed to write db secure boot variable: %r\n", err);
+                                uefi_call_wrapper(BS->Stall, 1, 3 * 1000 * 1000);
+                                return err;
+                        }
+
+                        Print(L"Finished loading secure boot variables. Rebooting.\n");
+                        uefi_call_wrapper(BS->Stall, 1, 3 * 1000 * 1000);
+
+                        err = uefi_call_wrapper(RT->ResetSystem, 4, EfiResetCold, EFI_SUCCESS, 0, NULL);
```

Comment:
hmm, is this really necessary? can't we go on?

i think most firmware leave setupmode anyway when the PK is set. I assumed we could go on then, without rebooting?

Reply:
Correct. SetupMode will be off so we don't need an explicit reset.

Reply:
Ok, I'll remove that part.

Although it is indeed part of the UEFI spec, I wonder how many firmware out there forget to leave setup mode after enrolling the PK.

Reply:
I think (hope?) this behaviour is correctly implemented as it's an explicit part of the spec and not really left as an exerciser for the reader to figure out. But who knows what shoddy firmware vendors does :) 

The best options is to verify it works with tianocore and consider everything else broken behavior.

Reply:
I can confirm it works just fine with tianocore.

Reply:
fixed